### DISCUSSION TITLE
storedCommand

### DISCUSSION THREAD

**Discussion by mskyttner** [proofrock/duckrg#1] [Ideas]
The storedStatement is nice, allowing "injection" of values into sql statements in a controlled way.

If running `duckrg` on the server, perhaps in a container, it would be possible to provide CLI commands for chains of piped commands that may mix nicely with (duck)sql (with read_json_auto, read_parquet, read_csv_auto etc)... 

So I was thinking of something similar to a storedStatement, like a storedCommand which could take parameters something along the lines of this (broken quotes, think "^C1" instead)...

```json
{
    "transaction": [
        {
            "statement": "install 'httpfs';"
	},
        {
            "statement": "load 'httpfs';"
	},
	{
            "query": "from read_json_auto('https://api.github.com/users/proofrock/events?per_page=100') select type, count(*) as event_count group by 1 order by 2 desc limit 10"
        },
	{
	    "command": "curl -sL 'https://api.github.com/users/##_username_##/events?per_page=100' \
    | duckdb -s 'COPY (SELECT type, count(*) AS event_count FROM read_json_auto('/dev/stdin') GROUP BY 1 ORDER BY 2 DESC LIMIT 10) TO '/dev/stdout' WITH (FORMAT 'csv', HEADER)' \
    | uplot bar -d, -H -t 'GitHub Events for @##_username_##'",
	    "values": { "username": "proofrock" }
	}
    ]
}
```

Using such storedCommands specified in the duckrg yaml would open it up to more than libduckdb sql returns but wouldn't be much different than storedStatements? 

Those pipes/chains/syscalls when used could either error out (just like sql statements) or return results with a content-type such as application/json or text/plain (or application/octet-stream?)

For example:

```bash
docker run --rm -v $(pwd)/gh_youplot.sh:/tmp/gh_youplot.sh ghcr.io/mskyttner/duckrg /tmp/gh_youplot.sh

                           GitHub Events for @dproofrock
                     ┌                                        ┐ 
           PushEvent ┤■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 51.0   
         CreateEvent ┤■■■■■■■■■■■■■■■ 22.0                      
         DeleteEvent ┤■■■■■■■■■■■ 17.0                          
        ReleaseEvent ┤■■■ 4.0                                   
         IssuesEvent ┤■ 2.0                                     
          WatchEvent ┤■ 2.0                                     
    PullRequestEvent ┤■ 1.0                                     
   IssueCommentEvent ┤■ 1.0                                     
                     └                                        ┘ 
```

(Here gh_youplot.sh is a bash script like in the yaml earlier ... but not taking any parameters)

**Comment by proofrock** [comment]
That's really interesting, thank you! 

For implementation, it could be easy/possible to start a subprocess. Some points to carefully design:

- parameters substitution, just `replaceAll` on a placeholder is maybe suboptimal, difficult to make not ambiguous without looking awkward. Also, different commands may need different escapes; I'd say that the client should take care of that.

- lifecycle management. What should stdin/stdout/stderr be "attached" to? How to intercept a fault?

- it's cool that the commands are in a transaction; how do they impact it? If the command fails, should I rollback? Also, transactions are not overlapping (there's a mutex), so if the command takes some time it blocks the database.

Then of course there's a portability issue. It should also work under windows, and in your example the pipes are implemented by the shell (`bash` usually) that it's not a windows'... direct... concept. You could of course just launch `bash -c` and pass everything to it, from the client, and that would be the best thing.

But since the client would do a bit of work, I question the usefulness of parameters, after all. In SQL, they're mainly there to avoid injection, because they map directly to a feature in the database. But here, it's arguably more work to properly escape the parameters than just composing the command with them inlined. The commonplace "tokenized" form for commands could be used:

```json
{
    "command": ["curl", "-sL", "https://api.github.com/users/proofrock/events?per_page=100", ...],
}
```

And it could be more manageable. What do you think? Thanks for now!
